File: src/lib/agent-bundle.ts
/**
 * Agent bundle loader + Tangle Sandbox `AgentProfile` translator.
 *
 * A "bundle" is a directory holding an `agent.json` plus the markdown system
 * prompts and methodology files referenced from it. The deploy script reads
 * the bundle, validates against `registry/_schemas/agent.schema.json`, then:
 *
 * 1. Translates the bundle-side shape into the SDK's portable `AgentProfile`
 *    (consumed as `backend.profile` by the sandbox runtime — used for the
 *    full-replacement system prompt, subagent profiles, model defaults, etc).
 * 2. Emits harness-native workspace files at `<workspace.root>` (default
 *    `/home/agent`): `AGENTS.md` (auto-loaded by every supported harness —
 *    OpenCode, Claude Code, Hermes, Codex, Amp, Kimi-Code), and — for
 *    multi-agent bundles only — `agents.json` (OpenCode subagent definitions).
 *    Plus every entry from `resources.files[]`.
 *
 * `AGENTS.md` and `agents.json` are auto-discovered by the in-sandbox harness
 * (see `apps/sidecar/src/agents/base-agent.ts:170` and
 * `apps/sidecar/src/agents/subagents/load-agents-config.ts`). The bundle's
 * job is to put them where the harness expects them.
 *
 * @public
 */
import { readFile, stat } from 'node:fs/promises'
import { dirname, isAbsolute, join, relative, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

// --- Types ----------------------------------------------------------------

/**
 * Permission policy for a runtime capability.
 * Mirrors `AgentProfilePermissionValue` in `@tangle-network/sandbox`.
 */
export type AgentPermission = 'allow' | 'ask' | 'deny'

/**
 * Default sandbox workspace root. Matches the harness convention
 * (`AGENT_WORKSPACE_ROOT=/home/agent` per
 * `apps/sidecar/src/constants.ts`).
 */
export const DEFAULT_WORKSPACE_ROOT = '/home/agent'

/**
 * A single resource file (or directory) to materialise inside the sandbox.
 */
export interface BundleResource {
  /** Path relative to the bundle root. May be a file or a directory. */
  source: string
  /**
   * Path inside the sandbox. If relative, treated as relative to the
   * resolved workspace root. Defaults to `<workspace.root>/<source>` when
   * omitted.
   */
  target?: string
}

/**
 * Bundle-side subagent declaration. Translates to:
 *  - `AgentProfile.subagents[id]` (SDK side)
 *  - one entry in the emitted `<workspace.root>/agents.json` (harness side)
 */
export interface SubagentBundle {
  description?: string
  systemPromptFile: string
  model?: string
  tools?: Record<string, boolean>
  permissions?: Record<string, AgentPermission>
  temperature?: number
  maxSteps?: number
}

/**
 * The complete bundle-side agent spec stored in `agent.json`.
 */
export interface AgentBundleProfile {
  name: string
  description?: string
  version: string
  tags?: string[]
  workspace?: {
    root?: string
  }
  prompt: {
    systemPromptFile?: string
    instructions?: string[]
  }
  model?: {
    preferred?: string
    fallback?: string[]
  }
  tools?: Record<string, boolean>
  permissions?: Record<string, AgentPermission>
  mcp?: Record<string, unknown>
  subagents?: Record<string, SubagentBundle>
  resources?: {
    files?: BundleResource[]
  }
}

// --- Translator target (AgentProfile mirror) ------------------------------

/**
 * Local mirror of the SDK's portable `AgentProfile`.
 *
 * We mirror — rather than import — the shape so this library has no hard
 * runtime dependency on `@tangle-network/sandbox`. The deploy script casts
 * the result to the SDK type at the call site; if the SDK shape drifts the
 * cast surfaces the issue at the boundary.
 *
 * @public
 */
export interface AgentProfileMirror {
  name?: string
  description?: string
  version?: string
  tags?: string[]
  prompt?: {
    systemPrompt?: string
    instructions?: string[]
  }
  model?: {
    default?: string
    metadata?: Record<string, unknown>
  }
  tools?: Record<string, boolean>
  permissions?: Record<string, AgentPermission>
  mcp?: Record<string, unknown>
  subagents?: Record<string, AgentSubagentProfileMirror>
  resources?: {
    files?: AgentProfileFileMountMirror[]
  }
  metadata?: Record<string, unknown>
}

export interface AgentSubagentProfileMirror {
  description?: string
  prompt?: string
  model?: string
  tools?: Record<string, boolean>
  permissions?: Record<string, AgentPermission>
  maxSteps?: number
}

export interface AgentProfileFileMountMirror {
  path: string
  resource:
    | { kind: 'inline'; name: string; content: string }
    | { kind: 'github'; path: string; ref?: string; name?: string }
  executable?: boolean
}

/**
 * One file the deploy script must `box.files.write` into the sandbox.
 *
 * Distinct from `AgentProfileFileMountMirror`: workspace files are the
 * harness-native artefacts (AGENTS.md, agents.json, methodology/*) emitted
 * verbatim — not declared on the AgentProfile.
 *
 * @public
 */
export interface WorkspaceFile {
  /** Absolute path inside the sandbox. */
  targetPath: string
  /** UTF-8 string content (binary file copies are not currently supported). */
  content: string
}

/**
 * Shape of a single entry in the OpenCode `agents.json` file.
 *
 * Verified against `apps/sidecar/agents.json` and
 * `apps/sidecar/src/agents/subagents/load-agents-config.ts`. Keys are the
 * harness contract — not invented here.
 *
 * @public
 */
export interface OpenCodeSubagentConfig {
  mode: 'subagent'
  description?: string
  prompt: string
  temperature?: number
  tools?: Record<string, boolean>
  permission?: Record<string, AgentPermission>
}

// --- Schema validation (minimal, matches scripts/validate-registry.ts) ---

interface SchemaNode {
  type?: string | string[]
  enum?: unknown[]
  required?: string[]
  properties?: Record<string, SchemaNode>
  additionalProperties?: boolean | SchemaNode
  items?: SchemaNode
  minimum?: number
  pattern?: string
}

function validateAgainst(value: unknown, schema: SchemaNode, path = '$'): string[] {
  const errors: string[] = []
  if (schema.type) {
    const expected = Array.isArray(schema.type) ? schema.type : [schema.type]
    const actual = Array.isArray(value) ? 'array' : value === null ? 'null' : typeof value
    if (!expected.includes(actual)) {
      errors.push(`${path}: expected ${expected.join('|')}, got ${actual}`)
      return errors
    }
  }
  if (schema.enum && !schema.enum.includes(value)) {
    errors.push(`${path}: ${JSON.stringify(value)} not in enum`)
  }
  if (schema.minimum !== undefined && typeof value === 'number' && value < schema.minimum) {
    errors.push(`${path}: ${value} < minimum ${schema.minimum}`)
  }
  if (schema.required && typeof value === 'object' && value !== null && !Array.isArray(value)) {
    for (const req of schema.required) {
      if (!(req in (value as Record<string, unknown>))) {
        errors.push(`${path}.${req}: required field missing`)
      }
    }
  }
  if (schema.properties && typeof value === 'object' && value !== null && !Array.isArray(value)) {
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      const propSchema = schema.properties[k]
      if (propSchema) {
        errors.push(...validateAgainst(v, propSchema, `${path}.${k}`))
      } else if (schema.additionalProperties === false) {
        errors.push(`${path}.${k}: unexpected property (additionalProperties=false)`)
      } else if (
        typeof schema.additionalProperties === 'object' &&
        schema.additionalProperties !== null
      ) {
        errors.push(...validateAgainst(v, schema.additionalProperties, `${path}.${k}`))
      }
    }
  }
  if (schema.items && Array.isArray(value)) {
    for (let i = 0; i < value.length; i++) {
      errors.push(...validateAgainst(value[i], schema.items, `${path}[${i}]`))
    }
  }
  return errors
}

let cachedSchema: SchemaNode | undefined
async function loadSchema(): Promise<SchemaNode> {
  if (cachedSchema) return cachedSchema
  // The schema lives at registry/_schemas/agent.schema.json relative to the
  // package root. Resolve from this module's own location so the loader is
  // usable from both src/ (tsx) and dist/ (built).
  const here = dirname(fileURLToPath(import.meta.url))
  // From src/lib/agent-bundle.ts → repo root is two up. From dist/lib/* → also two up.
  const schemaPath = resolve(here, '..', '..', 'registry', '_schemas', 'agent.schema.json')
  const raw = await readFile(schemaPath, 'utf8')
  cachedSchema = JSON.parse(raw) as SchemaNode
  return cachedSchema
}

// --- Loader ---------------------------------------------------------------

/**
 * Read `agent.json` from a bundle directory and validate it.
 *
 * @public
 * @throws when agent.json is missing, unparseable, or fails schema validation
 */
export async function loadAgentBundle(bundleDir: string): Promise<AgentBundleProfile> {
  const manifestPath = join(bundleDir, 'agent.json')
  let raw: string
  try {
    raw = await readFile(manifestPath, 'utf8')
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err)
    throw new Error(`agent.json not readable at ${manifestPath}: ${reason}`, { cause: err })
  }
  let parsed: unknown
  try {
    parsed = JSON.parse(raw)
  } catch (err) {
    const reason = err instanceof Error ? err.message : String(err)
    throw new Error(`agent.json at ${manifestPath} is not valid JSON: ${reason}`, { cause: err })
  }
  const schema = await loadSchema()
  const errors = validateAgainst(parsed, schema, 'agent')
  if (errors.length > 0) {
    throw new Error(
      `agent.json at ${manifestPath} failed schema validation:\n  ${errors.join('\n  ')}`,
    )
  }
  return parsed as AgentBundleProfile
}

/**
 * Read the system prompt content referenced by the bundle.
 *
 * Returns an empty string when the bundle declares no `systemPromptFile`.
 *
 * @public
 * @throws when the file is declared but cannot be read
 */
export async function resolveSystemPrompt(
  bundle: AgentBundleProfile,
  bundleDir: string,
): Promise<string> {
  const file = bundle.prompt.systemPromptFile
  if (!file) return ''
  const abs = isAbsolute(file) ? file : join(bundleDir, file)
  return readFile(abs, 'utf8')
}

/**
 * Resolve the workspace root from a bundle, falling back to the harness
 * default (`/home/agent`).
 *
 * @public
 */
export function resolveWorkspaceRoot(bundle: AgentBundleProfile): string {
  return bundle.workspace?.root ?? DEFAULT_WORKSPACE_ROOT
}

// --- Translator -----------------------------------------------------------

interface ResolvedFileMount {
  path: string
  source: string
  content: string
}

/**
 * Walk a `BundleResource` (file or directory) and produce one inline-resource
 * mount per file. Directories are walked recursively.
 *
 * Target paths normalise to absolute sandbox paths under the workspace root.
 */
async function resolveResource(
  resource: BundleResource,
  bundleDir: string,
  workspaceRoot: string,
): Promise<ResolvedFileMount[]> {
  const sourceAbs = isAbsolute(resource.source) ? resource.source : join(bundleDir, resource.source)
  const stats = await stat(sourceAbs)
  const targetRoot = resource.target
    ? isAbsolute(resource.target)
      ? resource.target
      : join(workspaceRoot, resource.target)
    : join(workspaceRoot, resource.source)

  if (stats.isFile()) {
    const content = await readFile(sourceAbs, 'utf8')
    return [{ path: targetRoot, source: sourceAbs, content }]
  }
  if (stats.isDirectory()) {
    const out: ResolvedFileMount[] = []
    const { readdir } = await import('node:fs/promises')
    const entries = await readdir(sourceAbs, { withFileTypes: true, recursive: true })
    for (const entry of entries) {
      if (!entry.isFile()) continue
      // Node 20.12+: parentPath; older fall back to path. Tolerate both.
      const parent =
        (entry as unknown as { parentPath?: string }).parentPath ??
        (entry as unknown as { path?: string }).path ??
        sourceAbs
      const fileAbs = join(parent, entry.name)
      const rel = relative(sourceAbs, fileAbs)
      const content = await readFile(fileAbs, 'utf8')
      out.push({
        path: join(targetRoot, rel),
        source: fileAbs,
        content,
      })
    }
    return out
  }
  throw new Error(`resource source ${sourceAbs} is neither file nor directory`)
}

/**
 * Translate an `AgentBundleProfile` into the SDK's portable `AgentProfile`.
 *
 * Inlines:
 * - `prompt.systemPromptFile` content into `prompt.systemPrompt`
 * - subagent `systemPromptFile` content into each `subagents[id].prompt`
 * - every `resources.files[]` entry (recursively, when source is a directory)
 *   as an inline file mount under the resolved sandbox target path
 *
 * @public
 */
export async function toAgentProfile(
  bundle: AgentBundleProfile,
  bundleDir: string,
): Promise<AgentProfileMirror> {
  const workspaceRoot = resolveWorkspaceRoot(bundle)
  const systemPrompt = await resolveSystemPrompt(bundle, bundleDir)

  const resolvedSubagents: Record<string, AgentSubagentProfileMirror> = {}
  if (bundle.subagents) {
    for (const [id, sub] of Object.entries(bundle.subagents)) {
      const promptAbs = isAbsolute(sub.systemPromptFile)
        ? sub.systemPromptFile
        : join(bundleDir, sub.systemPromptFile)
      const subPrompt = await readFile(promptAbs, 'utf8')
      resolvedSubagents[id] = {
        ...(sub.description !== undefined ? { description: sub.description } : {}),
        prompt: subPrompt,
        ...(sub.model !== undefined ? { model: sub.model } : {}),
        ...(sub.tools ? { tools: sub.tools } : {}),
        ...(sub.permissions ? { permissions: sub.permissions } : {}),
        ...(sub.maxSteps !== undefined ? { maxSteps: sub.maxSteps } : {}),
      }
    }
  }

  const fileMounts: AgentProfileFileMountMirror[] = []
  for (const resource of bundle.resources?.files ?? []) {
    const resolved = await resolveResource(resource, bundleDir, workspaceRoot)
    for (const mount of resolved) {
      const name = relative(bundleDir, mount.source) || mount.path
      fileMounts.push({
        path: mount.path,
        resource: { kind: 'inline', name, content: mount.content },
      })
    }
  }

  const profile: AgentProfileMirror = {
    name: bundle.name,
    ...(bundle.description !== undefined ? { description: bundle.description } : {}),
    version: bundle.version,
    ...(bundle.tags ? { tags: bundle.tags } : {}),
    ...(systemPrompt || bundle.prompt.instructions
      ? {
          prompt: {
            ...(systemPrompt ? { systemPrompt } : {}),
            ...(bundle.prompt.instructions ? { instructions: bundle.prompt.instructions } : {}),
          },
        }
      : {}),
    ...(bundle.model?.preferred || bundle.model?.fallback
      ? {
          model: {
            ...(bundle.model.preferred ? { default: bundle.model.preferred } : {}),
            ...(bundle.model.fallback ? { metadata: { fallback: bundle.model.fallback } } : {}),
          },
        }
      : {}),
    ...(bundle.tools ? { tools: bundle.tools } : {}),
    ...(bundle.permissions ? { permissions: bundle.permissions } : {}),
    ...(bundle.mcp ? { mcp: bundle.mcp } : {}),
    ...(Object.keys(resolvedSubagents).length > 0 ? { subagents: resolvedSubagents } : {}),
    ...(fileMounts.length > 0 ? { resources: { files: fileMounts } } : {}),
  }

  return profile
}

// --- Workspace-file emit (harness-native) ---------------------------------

/**
 * Build the OpenCode-shape `agents.json` content for a multi-agent bundle.
 *
 * Shape verified against `apps/sidecar/agents.json` (the canonical example
 * shipped with the sidecar) and `load-agents-config.ts`:
 *
 *   { <subagent-id>: { mode, description, prompt, temperature?, tools?, permission? } }
 *
 * The prompt is INLINE (full markdown content), not a path. `permission`
 * is singular per the harness contract, even though the bundle-side spec
 * uses plural `permissions` (we translate at the boundary).
 */
async function buildOpenCodeAgentsJson(
  bundle: AgentBundleProfile,
  bundleDir: string,
): Promise<string> {
  const out: Record<string, OpenCodeSubagentConfig> = {}
  for (const [id, sub] of Object.entries(bundle.subagents ?? {})) {
    const promptAbs = isAbsolute(sub.systemPromptFile)
      ? sub.systemPromptFile
      : join(bundleDir, sub.systemPromptFile)
    const promptContent = await readFile(promptAbs, 'utf8')
    const entry: OpenCodeSubagentConfig = {
      mode: 'subagent',
      ...(sub.description !== undefined ? { description: sub.description } : {}),
      prompt: promptContent,
      ...(sub.temperature !== undefined ? { temperature: sub.temperature } : {}),
      ...(sub.tools ? { tools: sub.tools } : {}),
      ...(sub.permissions ? { permission: sub.permissions } : {}),
    }
    out[id] = entry
  }
  return JSON.stringify(out, null, 2) + '\n'
}

/**
 * Compute the list of files the deploy script must `box.files.write`
 * into the sandbox at `<workspace.root>`:
 *
 * - `<workspace.root>/AGENTS.md`   — orchestrator system prompt
 *                                    (auto-loaded by the harness)
 * - `<workspace.root>/agents.json` — multi-agent bundles only;
 *                                    OpenCode subagent definitions
 *                                    (auto-loaded by `loadAgentsConfig`)
 * - One file per `resources.files[]` entry, target defaulting to
 *   `<workspace.root>/<source>` when not specified.
 *
 * Single-agent bundles do NOT emit `agents.json`. Their orchestrator
 * prompt is the AGENTS.md content (also pushed via the SDK
 * `AgentProfile.prompt.systemPrompt` for a full-replacement system
 * prompt — equivalent content, two delivery channels).
 *
 * @public
 */
export async function toWorkspaceFiles(
  bundle: AgentBundleProfile,
  bundleDir: string,
): Promise<WorkspaceFile[]> {
  const workspaceRoot = resolveWorkspaceRoot(bundle)
  const out: WorkspaceFile[] = []

  // 1. AGENTS.md — orchestrator system prompt at the workspace root
  const systemPromptContent = await resolveSystemPrompt(bundle, bundleDir)
  if (systemPromptContent) {
    out.push({
      targetPath: join(workspaceRoot, 'AGENTS.md'),
      content: systemPromptContent,
    })
  }

  // 2. agents.json — multi-agent bundles only
  if (bundle.subagents && Object.keys(bundle.subagents).length > 0) {
    const agentsJson = await buildOpenCodeAgentsJson(bundle, bundleDir)
    out.push({
      targetPath: join(workspaceRoot, 'agents.json'),
      content: agentsJson,
    })
  }

  // 3. resources.files[] — methodology/, README.md, role assets, etc.
  for (const resource of bundle.resources?.files ?? []) {
    const resolved = await resolveResource(resource, bundleDir, workspaceRoot)
    for (const mount of resolved) {
      out.push({ targetPath: mount.path, content: mount.content })
    }
  }

  return out
}

File: scripts/python/starter_foundry/agent_bundle.py
"""
Agent bundle loader + Tangle Sandbox AgentProfile translator (Python port).

1:1 functional port of `src/lib/agent-bundle.ts`. A "bundle" is a directory
holding an `agent.json` plus referenced markdown system-prompt and methodology
files. The deploy script reads the bundle, validates against
`registry/_schemas/agent.schema.json`, then:

1. Translates the bundle-side shape into the SDK's portable `AgentProfile`
   (consumed as `backend.profile` by the sandbox runtime — used for the
   full-replacement system prompt, subagent profiles, model defaults, etc).
2. Emits harness-native workspace files at `<workspace.root>` (default
   `/home/agent`): `AGENTS.md` (auto-loaded by every supported harness —
   OpenCode, Claude Code, Hermes, Codex, Amp, Kimi-Code), and — for
   multi-agent bundles only — `agents.json` (OpenCode subagent definitions).
   Plus every entry from `resources.files[]`.

`AGENTS.md` and `agents.json` are auto-discovered by the in-sandbox harness
(see `apps/sidecar/src/agents/base-agent.ts:170` and
`apps/sidecar/src/agents/subagents/load-agents-config.ts`).

Sandbox target paths are POSIX even when the host is Windows, so we use
`posixpath` (not `os.path`) for every path that lives inside the sandbox.
"""

from __future__ import annotations

import json
import os
import posixpath
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Literal

import jsonschema

# Permission policy for a runtime capability.
# Mirrors `AgentProfilePermissionValue` in `@tangle-network/sandbox`.
AgentPermission = Literal["allow", "ask", "deny"]

# Default sandbox workspace root. Matches the harness convention
# (`AGENT_WORKSPACE_ROOT=/home/agent` per `apps/sidecar/src/constants.ts`).
DEFAULT_WORKSPACE_ROOT = "/home/agent"


@dataclass(frozen=True)
class BundleResource:
    """A single resource file (or directory) to materialise inside the sandbox."""

    source: str
    target: str | None = None


@dataclass(frozen=True)
class SubagentBundle:
    """Bundle-side subagent declaration."""

    systemPromptFile: str
    description: str | None = None
    model: str | None = None
    tools: dict[str, bool] | None = None
    permissions: dict[str, AgentPermission] | None = None
    temperature: float | None = None
    maxSteps: int | None = None


@dataclass(frozen=True)
class AgentBundleProfile:
    """The complete bundle-side agent spec stored in `agent.json`."""

    name: str
    version: str
    prompt: dict[str, Any]
    description: str | None = None
    tags: list[str] | None = None
    workspace: dict[str, Any] | None = None
    model: dict[str, Any] | None = None
    tools: dict[str, bool] | None = None
    permissions: dict[str, AgentPermission] | None = None
    mcp: dict[str, Any] | None = None
    subagents: dict[str, SubagentBundle] | None = None
    resources: dict[str, Any] | None = None


@dataclass(frozen=True)
class WorkspaceFile:
    """One file the deploy script must `box.files.write` into the sandbox."""

    targetPath: str
    content: str


# ---------------------------------------------------------------------------
# Schema validation
# ---------------------------------------------------------------------------

_SCHEMA_CACHE: dict[str, Any] = {}


def _schema_path() -> Path:
    """Resolve the canonical `registry/_schemas/agent.schema.json`.

    The Python package lives at `scripts/python/starter_foundry/` so the repo
    root is three parents up. Honour `STARTER_FOUNDRY_REPO` for callers who
    install the package from a non-canonical location.
    """
    override = os.environ.get("STARTER_FOUNDRY_REPO")
    if override:
        candidate = Path(override) / "registry" / "_schemas" / "agent.schema.json"
        if candidate.is_file():
            return candidate
    here = Path(__file__).resolve().parent
    candidate = here.parent.parent.parent / "registry" / "_schemas" / "agent.schema.json"
    return candidate


def _load_schema() -> dict[str, Any]:
    if "schema" in _SCHEMA_CACHE:
        return _SCHEMA_CACHE["schema"]
    path = _schema_path()
    with path.open("r", encoding="utf-8") as fh:
        schema = json.load(fh)
    _SCHEMA_CACHE["schema"] = schema
    return schema


def _format_validation_error(err: jsonschema.ValidationError) -> str:
    """Mirror the TS validator's error shape:

    `agent.foo: required field missing`
    `agent.bar: 1 < minimum 2`
    `agent.tags[0]: expected string, got number`
    """
    path_parts = ["agent", *list(err.absolute_path)]
    if err.validator == "required":
        # message: "'foo' is a required property" → infer the missing field
        missing = err.message.split("'")[1] if "'" in err.message else err.message
        path = ".".join(str(p) for p in path_parts) + "." + missing
        return f"{path}: required field missing"
    path = path_parts[0]
    for part in path_parts[1:]:
        if isinstance(part, int):
            path = f"{path}[{part}]"
        else:
            path = f"{path}.{part}"
    return f"{path}: {err.message}"


# ---------------------------------------------------------------------------
# Loader
# ---------------------------------------------------------------------------


def _coerce_subagents(raw: Any) -> dict[str, SubagentBundle] | None:
    if not isinstance(raw, dict):
        return None
    out: dict[str, SubagentBundle] = {}
    for sid, sval in raw.items():
        out[sid] = SubagentBundle(
            systemPromptFile=sval["systemPromptFile"],
            description=sval.get("description"),
            model=sval.get("model"),
            tools=sval.get("tools"),
            permissions=sval.get("permissions"),
            temperature=sval.get("temperature"),
            maxSteps=sval.get("maxSteps"),
        )
    return out


def load_agent_bundle(bundle_dir: str | os.PathLike[str]) -> AgentBundleProfile:
    """Read `agent.json` from a bundle directory and validate it.

    Raises:
        FileNotFoundError-flavoured RuntimeError when agent.json is missing.
        ValueError when the JSON is unparseable.
        ValueError when schema validation fails (multi-line, like the TS port).
    """
    bundle_path = Path(bundle_dir)
    manifest_path = bundle_path / "agent.json"
    try:
        raw = manifest_path.read_text(encoding="utf-8")
    except OSError as err:
        raise ValueError(
            f"agent.json not readable at {manifest_path}: {err}"
        ) from err
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as err:
        raise ValueError(
            f"agent.json at {manifest_path} is not valid JSON: {err}"
        ) from err

    schema = _load_schema()
    validator = jsonschema.Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(parsed), key=lambda e: list(e.absolute_path))
    if errors:
        formatted = "\n  ".join(_format_validation_error(e) for e in errors)
        raise ValueError(
            f"agent.json at {manifest_path} failed schema validation:\n  {formatted}"
        )

    return AgentBundleProfile(
        name=parsed["name"],
        version=parsed["version"],
        prompt=parsed["prompt"],
        description=parsed.get("description"),
        tags=parsed.get("tags"),
        workspace=parsed.get("workspace"),
        model=parsed.get("model"),
        tools=parsed.get("tools"),
        permissions=parsed.get("permissions"),
        mcp=parsed.get("mcp"),
        subagents=_coerce_subagents(parsed.get("subagents")),
        resources=parsed.get("resources"),
    )


def resolve_system_prompt(bundle: AgentBundleProfile, bundle_dir: str | os.PathLike[str]) -> str:
    """Read the system prompt content referenced by the bundle.

    Returns the empty string when the bundle declares no `systemPromptFile`.
    """
    file = bundle.prompt.get("systemPromptFile")
    if not file:
        return ""
    p = Path(file)
    if not p.is_absolute():
        p = Path(bundle_dir) / file
    return p.read_text(encoding="utf-8")


def resolve_workspace_root(bundle: AgentBundleProfile) -> str:
    """Resolve the workspace root, falling back to /home/agent."""
    if bundle.workspace and bundle.workspace.get("root"):
        return bundle.workspace["root"]
    return DEFAULT_WORKSPACE_ROOT


# ---------------------------------------------------------------------------
# Resource resolution
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class _ResolvedFileMount:
    path: str
    source: str
    content: str


def _resolve_resource(
    resource: BundleResource | dict[str, Any],
    bundle_dir: str | os.PathLike[str],
    workspace_root: str,
) -> list[_ResolvedFileMount]:
    """Walk a `BundleResource` (file or directory) and produce inline mounts."""
    if isinstance(resource, dict):
        source = resource["source"]
        target = resource.get("target")
    else:
        source = resource.source
        target = resource.target

    bundle_path = Path(bundle_dir)
    source_path = Path(source) if Path(source).is_absolute() else bundle_path / source

    if target is None:
        target_root = posixpath.join(workspace_root, source)
    elif posixpath.isabs(target) or target.startswith("/"):
        target_root = target
    else:
        target_root = posixpath.join(workspace_root, target)

    if source_path.is_file():
        content = source_path.read_text(encoding="utf-8")
        return [
            _ResolvedFileMount(
                path=target_root,
                source=str(source_path),
                content=content,
            )
        ]
    if source_path.is_dir():
        out: list[_ResolvedFileMount] = []
        for entry in sorted(source_path.rglob("*")):
            if not entry.is_file():
                continue
            rel = entry.relative_to(source_path).as_posix()
            content = entry.read_text(encoding="utf-8")
            out.append(
                _ResolvedFileMount(
                    path=posixpath.join(target_root, rel),
                    source=str(entry),
                    content=content,
                )
            )
        return out
    raise ValueError(f"resource source {source_path} is neither file nor directory")


# ---------------------------------------------------------------------------
# Translator: bundle -> AgentProfile (the SDK's `backend.profile`)
# ---------------------------------------------------------------------------


def to_agent_profile(
    bundle: AgentBundleProfile,
    bundle_dir: str | os.PathLike[str],
) -> dict[str, Any]:
    """Translate an `AgentBundleProfile` into the SDK's portable AgentProfile.

    Inlines:
    - `prompt.systemPromptFile` content into `prompt.systemPrompt`
    - subagent `systemPromptFile` content into each `subagents[id].prompt`
    - every `resources.files[]` entry as an inline file mount under the
      resolved sandbox target path
    """
    workspace_root = resolve_workspace_root(bundle)
    system_prompt = resolve_system_prompt(bundle, bundle_dir)

    resolved_subagents: dict[str, dict[str, Any]] = {}
    if bundle.subagents:
        for sid, sub in bundle.subagents.items():
            prompt_path = Path(sub.systemPromptFile)
            if not prompt_path.is_absolute():
                prompt_path = Path(bundle_dir) / sub.systemPromptFile
            sub_prompt = prompt_path.read_text(encoding="utf-8")
            entry: dict[str, Any] = {}
            if sub.description is not None:
                entry["description"] = sub.description
            entry["prompt"] = sub_prompt
            if sub.model is not None:
                entry["model"] = sub.model
            if sub.tools:
                entry["tools"] = sub.tools
            if sub.permissions:
                entry["permissions"] = sub.permissions
            if sub.maxSteps is not None:
                entry["maxSteps"] = sub.maxSteps
            resolved_subagents[sid] = entry

    file_mounts: list[dict[str, Any]] = []
    bundle_path = Path(bundle_dir)
    for raw in (bundle.resources or {}).get("files", []) or []:
        resolved = _resolve_resource(raw, bundle_dir, workspace_root)
        for mount in resolved:
            try:
                rel = os.path.relpath(mount.source, bundle_path)
            except ValueError:
                rel = ""
            name = rel if rel else mount.path
            # Match the TS Node `path.relative` behaviour: forward slashes.
            name = name.replace(os.sep, "/")
            file_mounts.append(
                {
                    "path": mount.path,
                    "resource": {
                        "kind": "inline",
                        "name": name,
                        "content": mount.content,
                    },
                }
            )

    profile: dict[str, Any] = {"name": bundle.name}
    if bundle.description is not None:
        profile["description"] = bundle.description
    profile["version"] = bundle.version
    if bundle.tags:
        profile["tags"] = bundle.tags
    instructions = bundle.prompt.get("instructions")
    if system_prompt or instructions:
        prompt_obj: dict[str, Any] = {}
        if system_prompt:
            prompt_obj["systemPrompt"] = system_prompt
        if instructions:
            prompt_obj["instructions"] = instructions
        profile["prompt"] = prompt_obj
    if bundle.model and (bundle.model.get("preferred") or bundle.model.get("fallback")):
        model_obj: dict[str, Any] = {}
        if bundle.model.get("preferred"):
            model_obj["default"] = bundle.model["preferred"]
        if bundle.model.get("fallback"):
            model_obj["metadata"] = {"fallback": bundle.model["fallback"]}
        profile["model"] = model_obj
    if bundle.tools:
        profile["tools"] = bundle.tools
    if bundle.permissions:
        profile["permissions"] = bundle.permissions
    if bundle.mcp:
        profile["mcp"] = bundle.mcp
    if resolved_subagents:
        profile["subagents"] = resolved_subagents
    if file_mounts:
        profile["resources"] = {"files": file_mounts}

    return profile


# ---------------------------------------------------------------------------
# Workspace-file emit (harness-native AGENTS.md + agents.json + resources)
# ---------------------------------------------------------------------------


def _build_opencode_agents_json(
    bundle: AgentBundleProfile,
    bundle_dir: str | os.PathLike[str],
) -> str:
    """Build the OpenCode-shape `agents.json` content for a multi-agent bundle.

    Per `apps/sidecar/agents.json` + `load-agents-config.ts`:
        { <id>: { mode, description?, prompt, temperature?, tools?, permission? } }

    `permission` is SINGULAR (harness contract); the bundle-side spec uses
    plural `permissions` and we translate at the boundary.
    """
    out: dict[str, dict[str, Any]] = {}
    for sid, sub in (bundle.subagents or {}).items():
        prompt_path = Path(sub.systemPromptFile)
        if not prompt_path.is_absolute():
            prompt_path = Path(bundle_dir) / sub.systemPromptFile
        prompt_content = prompt_path.read_text(encoding="utf-8")
        entry: dict[str, Any] = {"mode": "subagent"}
        if sub.description is not None:
            entry["description"] = sub.description
        entry["prompt"] = prompt_content
        if sub.temperature is not None:
            entry["temperature"] = sub.temperature
        if sub.tools:
            entry["tools"] = sub.tools
        if sub.permissions:
            entry["permission"] = sub.permissions
        out[sid] = entry
    # ensure_ascii=False so non-ASCII content (em dashes, unicode arrows) lands
    # in agents.json verbatim — matches the TS port's JSON.stringify output.
    return json.dumps(out, indent=2, ensure_ascii=False) + "\n"


def to_workspace_files(
    bundle: AgentBundleProfile,
    bundle_dir: str | os.PathLike[str],
) -> list[WorkspaceFile]:
    """Compute the list of files the deploy script must `box.files.write`.

    - `<workspace.root>/AGENTS.md`   — orchestrator system prompt (always, when the
                                       bundle declares a `prompt.systemPromptFile`)
    - `<workspace.root>/agents.json` — multi-agent bundles only; OpenCode subagents
    - One file per `resources.files[]` entry, target defaulting to
      `<workspace.root>/<source>` when not specified.
    """
    workspace_root = resolve_workspace_root(bundle)
    out: list[WorkspaceFile] = []

    # 1. AGENTS.md — orchestrator system prompt at the workspace root
    system_prompt_content = resolve_system_prompt(bundle, bundle_dir)
    if system_prompt_content:
        out.append(
            WorkspaceFile(
                targetPath=posixpath.join(workspace_root, "AGENTS.md"),
                content=system_prompt_content,
            )
        )

    # 2. agents.json — multi-agent bundles only
    if bundle.subagents and len(bundle.subagents) > 0:
        agents_json = _build_opencode_agents_json(bundle, bundle_dir)
        out.append(
            WorkspaceFile(
                targetPath=posixpath.join(workspace_root, "agents.json"),
                content=agents_json,
            )
        )

    # 3. resources.files[]
    for raw in (bundle.resources or {}).get("files", []) or []:
        resolved = _resolve_resource(raw, bundle_dir, workspace_root)
        for mount in resolved:
            out.append(WorkspaceFile(targetPath=mount.path, content=mount.content))

    return out


# Silence unused-warning on the `field` import: kept available for future
# dataclass extensions without re-importing.
_ = field
